docs/reinterpreting_effects.md

```diff
@@ -0,0 +1,359 @@
+
+Let's take a moment to think about what could go wrong with an HTTP API from which we plan to fetch some JSON and convert it into a list of `CatFact`s.
+
+We can conceive that the server might occasionally return a malformed JSON response:
+
+``` haskell
+newtype JsonParseError = JsonParseError String
+  deriving (Show, Eq)
+
+decodeOrThrow :: (Has (Throw JsonParseError) sig m, FromJSON a) => L.ByteString -> m a
+decodeOrThrow = either (throwError . JsonParseError) pure . eitherDecode
+```
+
+A more HTTP-centric issue is that we might receive a content type we can't use. In this case, anything that's not `application/json`:
+
+``` haskell
+newtype InvalidContentType = InvalidContentType String
+  deriving (Show, Eq)
+
+```
+
+Now we need to support fetching JSON given an HTTP request. We have no guarantee that an arbitrary HTTP request
+will actually return JSON, so for this implementation we have to account for failure conditions. This provides
+a great opportunity to see how effect handlers can actually rely on *multiple underlying effects*!
+
+``` haskell
+newtype CatFactsApi m a = CatFactsApi { runCatFactsApi :: m a }
+ deriving newtype
+      ( Monad
+      , Functor
+      , Applicative
+      , MonadIO
+      , Alternative
+      )
+
+catFactsEndpoint :: HTTP.Request
+catFactsEndpoint = HTTP.parseRequest_ "https://cat-fact.herokuapp.com/facts/random"
+
+instance ( Has Http sig m
+         , Has (Throw JsonParseError) sig m
+         , Has (Throw InvalidContentType) sig m
+         , Algebra sig m
+         ) =>
+         Algebra (CatFactClient :+: sig) (CatFactsApi m) where
+  eff (L (ListFacts numberOfFacts k)) = do
+    resp <- sendRequest (catFactsEndpoint { HTTP.queryString = "?amount=" <> B.pack (show numberOfFacts) })
+    case lookup hContentType (HTTP.responseHeaders resp) of
+      Just "application/json; charset=utf-8" -> decodeOrThrow (HTTP.responseBody resp) >>= k
+      other -> throwError (InvalidContentType (show other))
+  eff (R other) = CatFactsApi (eff (handleCoercible other))
+```
+
+We implement a `CatFacts` effect handler that depends on _three_ underlying effects:
+
+1. `Http` - we need to be able to make requests
+2. `Throw JsonParseError` - we need to be able to signal that some aspect of the JSON wasn't what we expected.
+3. `Throw InvalidContentType` - we need to be able to signal what we got wasn't JSON at all!
+
+The nice aspect of this is that we have neatly contained the failure scenarios to their relevant strata rather than leaking them into the higher-level abstraction (`listFacts`)!
+
+Now we need to support performing HTTP requests:
+
+``` haskell
+
+newtype HttpClient m a = HttpClient { runHttp :: m a }
+ deriving newtype
+      ( Monad
+      , Functor
+      , Applicative
+      , MonadIO
+      , Alternative
+      )
+
+instance (MonadIO m, Algebra sig m) => Algebra (Http :+: sig) (HttpClient m) where
+  eff (L (SendRequest req k)) = liftIO (HTTP.getGlobalManager >>= HTTP.httpLbs req) >>= k
```

Comment:
```suggestion
  alg (L (SendRequest req k)) = liftIO (HTTP.getGlobalManager >>= HTTP.httpLbs req) >>= k
```

This is a great example of working at the “right” level of abstraction—lifting the HTTP actions into `HttpClient` vs. putting everything inside `HttpClient $ …` and therefore having to do `>>= runHttp . k` at the end there. I really dig this.